### PR TITLE
BUG/ENH: Tweeide parameterization and better handling of exposure, inclusion of 'perry' and 'taylor' methods.

### PR DESCRIPTION
**Pull request by thequackdaddy**
Hello,

See #2915 for the cause of this. 

(I just realized that you probably wanted me to name the method 'gamma-glm', but I named it 'perry'. I also made a 'taylor' which is supposed to be WLS version of the same thing. If you want me to rename, let me know.)

I'm not sure this totally works. I first used the `fair` dataset to see how well/poorly this works (I generated sample weights, exposures. On a scale of 1-10, I'd give that a 7 as far as how good it felt. `taylor` and `perry` were very close, but the rest was quite different. 

When I ran this on my real dataset, my confidence fell to a 2. I'm getting much different estimates for p using all the different methods. So I might need to think this out more.

```python
In [4]: model = sm.GLM(
   ...:     endog=np.array(data['PDIncurredLoss']),
   ...:     exog=bx,
   ...:     family=sm.families.Tweedie(var_power=1.639, link_power=0),
   ...:     exposure=np.array(data['PDExposure'])
   ...: )
   ...: 
   ...: res = model.fit(rtol=1e-2, tol_criterion='params', maxiter=10)
   ...: p = model.estimate_tweedie_power(res.mu)

In [5]: model.estimate_tweedie_power(res.mu, method='taylor')
Out[5]: 2.0409433864515334

In [6]: model.estimate_tweedie_power(res.mu, method='perry')
Out[6]: 1.1050768778823175

In [7]: model.estimate_tweedie_power(res.mu, method='brentq')
Out[7]: 1.6455822161384193

In [8]: res.summary()
Out[8]: 
<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                      y   No. Observations:              3569086
Model:                            GLM   Df Residuals:                  3569058
Model Family:                 Tweedie   Df Model:                           27
Link Function:                    log   Scale:                   2553.65026079
Method:                          IRLS   Log-Likelihood:                    nan
Date:                Sun, 26 Jun 2016   Deviance:                   7.7973e+07
Time:                        20:20:35   Pearson chi2:                 9.11e+09
No. Iterations:                    10                                         
```

On the bright side, all the methods feel like converge with a guess-and-check approach. So that keeps it above a 1.

### CODE REVIEW

File: statsmodels/genmod/generalized_linear_model.py [hunk 1]
```diff
@@ -700,18 +713,82 @@ def estimate_tweedie_power(self, mu, method='brentq', low=1.01, high=5.):
...
+        if exposure is not None and not isinstance(self.family.link,
+                                                   families.links.Log):
+            raise ValueError("exposure can only be used with the log link function")
+
+        # Use fit exposure if appropriate
+        if exposure is None and hasattr(self, 'exposure'):
+            # Already logged
+            exposure = np.exp(self.exposure)
+        elif exposure is None:
+            exposure = np.ones(len(self.endog))
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: I think `exposure = 1.` should work correctly with broadcasting in the following.

File: statsmodels/genmod/generalized_linear_model.py [hunk 2]
```diff
@@ -700,18 +713,82 @@ def estimate_tweedie_power(self, mu, method='brentq', low=1.01, high=5.):
...
+
+        # Use fit freq_weights if appropriate
+        if freq_weights is None and hasattr(self, 'freq_weights'):
+            # Already logged
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: comment doesn't apply

File: statsmodels/genmod/generalized_linear_model.py [hunk 3]
```diff
@@ -700,18 +713,82 @@ def estimate_tweedie_power(self, mu, method='brentq', low=1.01, high=5.):
...
+            freq_weights = self.freq_weights
+        elif exposure is None:
+            freq_weights = np.ones(len(self.endog))
+
+        weights = freq_weights * exposure
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: I think there shouldn't be the exposure here. just weightss=freq_weights

**Comment by josef-pkt** [comment]
If there are descriptive names available, then we try to avoid using author names.
I think perry -> glm-gamma and taylor -> ols (or ols-logs)   
wls for weights is just and extension/variation on ols

File: statsmodels/genmod/generalized_linear_model.py [hunk 4]
```diff
@@ -700,18 +713,82 @@ def estimate_tweedie_power(self, mu, method='brentq', low=1.01, high=5.):
...
+
+            if both:
+                return np.exp(scale), power
+            else:
+                return power
+
+        elif method == 'taylor':
+            from statsmodels.regression.linear_model import WLS
+
+            y = np.log(np.power(endog2 - mu2, 2))
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: my guess is that we need to clip `endog2 - mu2` to be larger than zero. Somebody will come up with a case with zero residuals. I'm not sure what's a good threshold, 1e-6?

File: statsmodels/genmod/generalized_linear_model.py (hunk 3 above)
Comment (thequackdaddy) [inline statsmodels/genmod/generalized_linear_model.py]: So I struggle with this dilemma. 

Reading the Tweedie papers, it seems to me that they assume that `endog` is a rate per exposure... so they divide `endog` by `exposure` and use that as the new `endog`. Then they have `var_weights` which they set equal to exposure. `params` gives you the same value, but as we've discussed, this gives you different `cov_params` and the like as the `scale` variable is different. 

So in the `brentq` calculation, I'm calculating chi2 the same way as we do for a typical GLM... unadjusted for exposure. However the psi_p part, I am adjusting for exposure. I think what happens if I have a record that represents a policy that was in force for 1 day and there is a loss, you get a massive, massive endog value. (A policy in force for one day with a $5,000 claim has the same rate per year as an $18.25 million loss. I think we want to assign less weight to this scenario.)

As a side note, I was able to to run this function successfully on my 2 largest datasets, and both provided values of around 1.6, which is about what the literature says they should be. Those values also match the implied values from the alpha parameter that I derived for the gamma distribution of the loss.

Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: If we don't have reference cases to verify that the code/formulas are correct, then we need to go back to simulating tweedie distributions and check the results, and possibly do a small monte carlo.

I am not able to check the correctness untill I have a large enough block of time to go through formula derivation again (and my notes are in Montreal).

my guess: exposure should automatically downweight observations with small exposure and we shouldn't need to do it explicitly through the weights. I think chi2/dispersion will also need to be adjusted for exposure.

In your 1-day example the downweighting should occur through exposure in our case, and through small var_weights (large variance) in the case where endog is divided by exposure. (unless I have a basic mistake in my thinking.)

Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: It's likely that what you have is correct, but it's not written in a way how it would follow from an explicit derivation of the formulas. 
The variance of uncorrected endog should be linear in exposure with exposure corrected mu in the power mu2**p.  var(y) = exposure * s2 * mu2**p   where s2 and mu2 are scale and mean at exposure=1, i.e. per unit of exposure.

**Comment by thequackdaddy** [comment]
I put together a gist of what I *think* you mean by a monte-carlo simulation. 

The implied p that I'd like to calculate is ~1.76, however, I'm getting ~1.95. 

https://gist.github.com/thequackdaddy/2e4690a3eb8964b207d5122d41c7e7f4

Didn't spend a lot of time on this, so could be a problem lurking somewhere. The p-values are very, very low, so perhaps I need to make the model not as good?

**Comment by josef-pkt** [comment]
@thequackdaddy  I have not yet tried to catch up with this and related issues.

Is there a bug somewhere that should be fixed for statsmodels 0.8, or can this wait for 0.9?

**Comment by thequackdaddy** [comment]
@josef-pkt Sorry for not getting back to you earlier. I'm working on some work things and haven't had time to commit anything. 

I think there is a bug if you are working with `exposure`. If you aren't using exposure weighting, I think it kind of works. 

I'm happy to have this wait until 0.9. I wrote a Notebook, by the way, that uses the series approximation method (like the R dtweedie package). It works very well and runs reasonably quickly on my datasets. Unfortunately, that's on my work laptop. I'll try to clean it up and upload it as a gist.  

So feel free to leave as-is and we'll clean up later. If you have free time to review the notebook once its ready, I'd appreciate it. I really need someone much smarter than me to buy-in before I believe what I've done myself. 

On a side note, the Tweedie functionality that we've programmed so far works very well for my models. I'm hoping to have management approve gen1 in the next few weeks with production following a few months later. I've basically assumed a Tweedie variance power parameter of 1.6 based on the CV of the size of loss distributions. The models validate extremely well and far, far outperform our programs, so that's exciting. The bad news (from my perspective) is that with significant performance gains, that means many customers will see large increases (and some will see decreases), so management will have some heartburn from that. That's more an implementation question than a model question, however. 

My department has increased staffing and after some new people are trained I'm hoping maybe I can be more active.

**Comment by thequackdaddy** [comment]
After experimenting with this some more, I'm really now leaning towards thinking that the BrentQ/psi method might only really work in some limited situations. 

[I made this gist)[https://gist.github.com/thequackdaddy/25a32272e1219973eab308948ab89a2a] showing the two methods on a random dataset I generated. The SeriesApproximation method works extremely well--the optimum p-values from the scipy `minimize` function is very close to the actual p-value. 

The psi_p value is all over the map. 

@josef-pkt - At one point, you suggested that you were able to get the psi_p method to approxmiate very well. Do you have that code? I'm wondering if I've screwed something up. (Not unlikely).

**Comment by josef-pkt** [comment]
AFAIR, using the GLM-gamma worked well.
Using brentq worked well for me only in larger samples.
In small samples (100 nobs or something like this) the psi function was not monotonic, and IIRC had several zeros/roots. (It might be related to solving the scale inside the psi function.)

(I'm trying to finish release preparation, and might not look closely very soon.)

**Comment by thequackdaddy** [comment]
@josef-pkt As mentioned in the other thread, I think this works pretty well. 

Would you have some time to work through this with me :)

**Comment by thequackdaddy** [comment]
Not really actively working on this anymore as I think its pretty much good enough for me for now. 

Comparing to R's tweedie package, there's also qtweedie and ptweedie which return the quantiles and cumulative distribution (Pr(X <= x)) which would be very useful for q-q plots. I can't seem to find literature on that though. Having that would allow me to run q-q plots--which I've never done on Tweedie but I found a stackexchange explanation of. Seems do-able...  good wishlist item.

**Comment by thequackdaddy** [comment]
Since exploring with `var_weights`, I think its fair to say that exposure requires some incorporation of `var_weights`. I removed all the exposure logic... For now, I'm just getting this to pass again and fixing a merge conflict.

**Comment by thequackdaddy** [comment]
I should add that I need to see how accurate/inaccurate this all is if there are `var_weights` as that is truly important to me. I'm not entirely sure at this point. I would imagine that `taylor` and `perry` won't work either (I'm not sure how variance weights would work in that case.)

File: statsmodels/genmod/generalized_linear_model.py [hunk 5]
```diff
@@ -784,18 +790,50 @@ def estimate_tweedie_power(self, mu, method='brentq', low=1.01, high=5.):
...
+            low = kwargs.get('low', 1.01)
+            high = kwargs.get('high', 1.99)
+
+            def difference(p):
+                scale = ((self.freq_weights * (self.endog - mu) ** 2 /
+                          mu ** p).sum() / self.df_resid)
+                scale /= self.var_weights
+
+                def psi_p(power):
+                    psi = ((self.endog - mu) ** 2 / (scale * mu ** power) - 1)
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: var_weights always needs to go in in the same way as inverse scale (1 / scale), i.e. line 806 above
so, this looks fine to me.

File: statsmodels/genmod/generalized_linear_model.py [hunk 6]
```diff
@@ -784,18 +790,50 @@ def estimate_tweedie_power(self, mu, method='brentq', low=1.01, high=5.):
...
+
+            power = minimize_scalar(difference, bounds=(low, high)).x
+
+        elif method == 'perry':
+            y = np.power(self.endog - mu, 2)
+            x = np.column_stack((np.ones(len(self.endog)), np.log(mu)))
+            model = GLM(y, x, freq_weights=weights,
+                        family=families.Gamma(families.links.log))
+            res = model.fit()
+            scale, power = res.params
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: res.params here is in log scale, i.e. links.log
Don't you need to rescale params? Or is this implied/correctly parameterized in model?

File: statsmodels/genmod/generalized_linear_model.py [hunk 7]
```diff
@@ -784,18 +790,50 @@ def estimate_tweedie_power(self, mu, method='brentq', low=1.01, high=5.):
...
+
+        elif method == 'taylor':
+            from statsmodels.regression.linear_model import WLS
+
+            y = np.log(np.power(self.endog - mu, 2))
+            x = np.column_stack((np.ones(len(self.endog)), np.log(mu)))
+            model = WLS(endog=y, exog=x, weights=weights)
+            res = model.fit()
+            scale, power = res.params
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: here there is no log-link, so I think either this or perry/glm has the wrong parameterization. 
?

Comment (josef-pkt) [reply statsmodels/genmod/generalized_linear_model.py]: correction: endog here is `log`ed

**Comment by josef-pkt** [comment]
I don't want to take the time right now to go back into the optimization problems here with the power parameter. The gist also has runtime warnings that we might be able to work around by clipping or restricting to valid ranges. (I didn't check, and don't remember running into it when I was playing with Tweedie.)

Given that this is just a helper method and won't directly affect other code, I think we can merge if you think it works better than the current version.
GLM-gamma and WLS, perry and taylor, methods look fine but I didn't try to check the math.

(Based on my earlier experiments, it might be necessary to use some global optimizer to work around multiple roots. I'm not sure about that because we don't have a loglike to rank roots/zeros.

Aside: Why does the precision of the var_power matter? Mean parameters or glm tweedie should be fine even if tweedie var power is off by a bit (?). Maybe we can use other, higher moments or quantiles or ... to better match the power parameter. E.g. like trying to find the appropriate count distribution after we have more of them.)
